File: antigravity-backend/services/vector_memory.py
import os
from chromadb import Client
from chromadb.config import Settings

chroma_path = os.getenv("CHROMA_DB_PATH", "./data/vector_db")
client = Client(Settings(persist_directory=chroma_path))
collection_name = "antigravity_memory"


def save_memory(id: str, text: str):
    collection = client.get_or_create_collection(collection_name)
    collection.add(documents=[text], ids=[id])


def search_memory(query: str, n_results: int = 5):
    collection = client.get_or_create_collection(collection_name)
    results = collection.query(query_texts=[query], n_results=n_results)
    return {
        "ids": results.get("ids", [[]])[0],
        "documents": results.get("documents", [[]])[0],
        "distances": results.get("distances", [[]])[0]
    }


def delete_memory(id: str):
    collection = client.get_or_create_collection(collection_name)
    collection.delete(ids=[id])


def clear_memory():
    client.delete_collection(collection_name)

File: services/vector_memory.ts
import { Client } from "chromadb";

let client: Client | null = null;
let collectionName = "antigravity_memory";

function getClient(): Client {
  if (!client) {
    client = new Client({
      path: "http://localhost:8000",
    });
  }
  return client;
}

export async function saveMemory(id: string, text: string): Promise<void> {
  const chroma = getClient();
  const collection = await chroma.getOrCreateCollection({
    name: collectionName,
  });

  await collection.add({
    documents: [text],
    ids: [id],
  });
}

export async function searchMemory(
  query: string,
  nResults: number = 5
): Promise<{ ids: string[]; documents: string[]; distances: number[] }> {
  const chroma = getClient();
  const collection = await chroma.getOrCreateCollection({
    name: collectionName,
  });

  const results = await collection.query({
    queryTexts: [query],
    nResults,
  });

  return {
    ids: results.ids[0] || [],
    documents: results.documents[0] || [],
    distances: results.distances?.[0] || [],
  };
}

export async function deleteMemory(id: string): Promise<void> {
  const chroma = getClient();
  const collection = await chroma.getOrCreateCollection({
    name: collectionName,
  });

  await collection.delete({ ids: [id] });
}

export interface MemorySearchResult {
  ids: string[];
  documents: string[];
  distances: number[];
}

export async function clearMemory(): Promise<void> {
  const chroma = getClient();
  await chroma.deleteCollection({ name: collectionName });
}